crates/matrix-sdk-crypto/src/identities/room_identity_state.rs

```diff
@@ -0,0 +1,1078 @@
+            }
+        }
+
+        // We didn't find a relevant update, so return an empty list
+        vec![]
+    }
+
+    fn update_user_state(
+        &mut self,
+        user_id: &UserId,
+        user_identity: &UserIdentity,
+    ) -> Option<IdentityStatusChange> {
+        if let UserIdentity::Other(_) = &user_identity {
+            // If the user's state has changed
+            let new_state = state_of(user_identity);
+            let old_state = self.known_states.get(user_id);
+            if new_state != old_state {
+                Some(self.set_state(user_identity.user_id(), new_state))
+            } else {
+                // Nothing changed
+                None
+            }
+        } else {
+            // Ignore updates to our own identity
+            None
+        }
+    }
+
+    fn set_state(&mut self, user_id: &UserId, new_state: IdentityState) -> IdentityStatusChange {
+        // Remember the new state of the user
+        self.known_states.set(user_id, &new_state);
+
+        // And return the update
+        IdentityStatusChange { user_id: user_id.to_owned(), changed_to: new_state }
+    }
+}
+
+fn state_of(user_identity: &UserIdentity) -> IdentityState {
+    if user_identity.is_verified() {
+        IdentityState::Verified
+    } else if user_identity.has_verification_violation() {
+        IdentityState::PreviouslyVerified
+    } else if let UserIdentity::Other(u) = user_identity {
+        if u.identity_needs_user_approval() {
+            IdentityState::PinViolation
+        } else {
+            IdentityState::Pinned
+        }
+    } else {
+        IdentityState::Pinned
+    }
+}
+
+/// A change in the status of the identity of a member of the room. Returned by
+/// [`RoomIdentityState::process_change`] to indicate that something changed in
+/// this room and we should either show or hide a warning.
+#[derive(Clone, Debug, PartialEq)]
+pub struct IdentityStatusChange {
+    /// The user ID of the user whose identity status changed
+    pub user_id: OwnedUserId,
+
+    /// The new state of the identity of the user
+    pub changed_to: IdentityState,
+}
+
+/// The state of an identity - verified, pinned etc.
+#[derive(Clone, Debug, PartialEq)]
+pub enum IdentityState {
+    /// The user is verified with us
+    Verified,
+
+    /// Either this is the first identity we have seen for this user, or the
+    /// user has acknowledged a change of identity explicitly e.g. by
+    /// clicking OK on a notification.
+    Pinned,
+
+    /// The user's identity has changed since it was pinned. The user should be
+    /// notified about this and given the opportunity to acknowledge the
+    /// change, which will make the new identity pinned.
+    PinViolation,
+
+    /// The user's identity has changed, and before that it was verified. This
+    /// is a serious problem. The user can either verify again to make this
+    /// identity verified, or withdraw verification to make it pinned.
+    PreviouslyVerified,
```

Comment:
As said previously, I'd keep `VerificationViolation` to be consistent with the identity api name?